### ISSUE TITLE
Seeking help making differentiable look-up table

### ISSUE THREAD

**Issue by s1ddok** [tensorflow/swift-apis#815]
I'm not sure if this is an issue with `swift-api` but rather a question I don't know where to ask elsewhere. What I'm trying to achieve is learnable Look-Up table. So basically what I want to do is to map every element in my tensor to an index (currently by just casting Float to Int) and then map that index to a learnable parameter inside my LUT. However I can't achieve that, I tried different approaches, this is the most straightforward one: 

```swift
import TensorFlow

typealias Tensorf = Tensor<Float>

let input = Tensorf(ones: [1, 256, 256, 1])

struct LookUpTable: Layer {
    
    var table: Tensorf
    
    init(table: Tensorf) {
        self.table = table
    }
    
    @differentiable
    func callAsFunction(_ input: Tensorf) -> Tensorf {
        var mappedScalars = input.scalars.map { s -> Float in
            let index = withoutDerivative(at: Int(s))
            return table[index].scalarized()
        }
        
        return Tensorf(shape: input.shape, scalars: mappedScalars)
    }
}

let lut = LookUpTable(table: Tensorf(ones: [10]))

let output = lut(input)
```

I keep getting error `function is not differentiable`. Is it possible to implement?

**Comment by dan-zheng** [comment]
Could you use a `Dictionary` instead, or it is important for the lookup table to be a tensor?

@marcrasi looked into conforming `Dictionary` to `Differentiable`, so you could use `var table: [Int: Tensor<Float>]`.

**Comment by s1ddok** [comment]
@dan-zheng thanks for the suggestion! I think `Array` could work for me as well since elmenets can be indices, but unfortunately it still doesn't work, here is what I've tried:

```
struct LookUpTable: Layer {

    var table: [Tensorf]

    init(table: [Tensorf]) {
        self.table = table
    }

    @differentiable
    func callAsFunction(_ input: Tensorf) -> Tensorf {
        let mappedScalars = input.scalars.differentiableMap { s -> Float in
            let index = withoutDerivative(at: Int(s))
            return table[index].scalarized()
        }

        return Tensorf(shape: input.shape, scalars: mappedScalars)
    }
}
```

Still getting `function is not differentiable`

**Comment by s1ddok** [comment]
Few updates on the topic. 

I think LUTs aren't differentiable with nearest neighbor sampling. I think they are differentiable with bilinear one (PyTorch's grid_sample kind of proves that)

I've made a task easier for myself, now I still want LUT to be learnable, but I don't want gradients to come through it inputs I wrote a code which runs, but crashes in runtime with `EXC_BAD_ACCESS`.

Any ideas on how to debug this? 

```
import Foundation
import TensorFlow

typealias Tensorf = Tensor<Float>

let input = Tensorf(randomNormal: [1, 256, 256, 12])

let semanticMap = Tensorf(ones: [1, 256, 256, 1])


struct CLADE: Layer {

    struct Input: Differentiable {
        var input: Tensorf
        @noDerivative var semanticMap: Tensorf
    }

    var table: Tensorf

    init(table: Tensorf) {
        self.table = table
    }

    @differentiable
    func callAsFunction(_ input: Input) -> Tensorf {
        let smap = withoutDerivative(at: input.semanticMap.scalars) {
            $0.map { Int($0) }
        }
        
        var scalars: [Tensorf] = []
        for element in smap {
            scalars.append(self.table[element])
        }

        let gammaTensor = Tensorf(stacking: scalars).reshaped(to: input.semanticMap.shape)
        
        return input.input * gammaTensor
    }
}

var clade = CLADE(table: .init([0.6, 0.8]))

var optimizer = Adam(for: clade)

while true {
    let cladeGrad = gradient(at: clade) { c -> Tensorf in
        let output = c(.init(input: input, semanticMap: semanticMap))
        
        return output.sum()
    }
    
    optimizer.update(&clade, along: cladeGrad)
    
    print(clade.table)
}
```

**Comment by marcrasi** [comment]
That code compiles, runs, and prints out some tensors for me on Linux with the S4TF v0.8 toolchain. What OS and toolchain version are you using?

Have you seen the [Embedding layer](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Layers/Embedding.swift)? It seems like it does something very similar to what you are doing, except that it takes integers as inputs instead of floats. Could you achieve your goal by writing a layer that casts its inputs to Int and then passes them through an embedding layer?

> I think LUTs aren't differentiable with nearest neighbor sampling. I think they are differentiable with bilinear one (PyTorch's grid_sample kind of proves that)
> 
> I've made a task easier for myself, now I still want LUT to be learnable, but I don't want gradients to come through it inputs

If you do eventually want to try to get the gradients to come through the inputs, I think you should be able to get it to work by interpolating the results of the embedding layer. Something like this (warning, I have compiled this but I have not run it to see if it gives reasonable results):

```swift
struct InterpolatedEmbedding: Layer {
  var embedding: Embedding<Float>

  @differentiable
  func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
    // input = integerPart + fractionalPart, where
    //   integerPart is an integer
    //   fractionalPart is in [0, 1)
    let integerPart = floor(input)
    let fractionalPart = input - integerPart

    let embedding1 = withoutDerivative(at: embedding(Tensor<Int32>(integerPart)))
    let embedding2 = withoutDerivative(at: embedding(Tensor<Int32>(integerPart) + 1))
    return (1 - fractionalPart) * embedding1 + fractionalPart * embedding2
  }
}
```

**Comment by marcrasi** [comment]
(Note: I edited my above comment because it originally had `withoutDerivative(at:` in the wrong place. Make sure to use the new verison.)

**Comment by s1ddok** [comment]
@marcrasi thanks for your replies!

I tried running the code on remote Linux server with CUDA 10.1 and v0.8 toolchain and it really appears to work. Unfortunately this is veeeeery slow, I will try to look at it. However it still does not work on nightly macOS toolchain with latest Xcode. 

Thanks for pointing me at `Embedding` thing, I will definitely try to look at it!

**Comment by s1ddok** [comment]
I've also tried to optimize stuff by using Floats and getting really weird error on Linux:

```
    @differentiable
    func callAsFunction(_ input: Input) -> Tensorf {
        let smap = withoutDerivative(at: input.semanticMap.scalars) {
            $0.map { Int($0) }
        }
        
        var scalars: [Float] = []
        for element in smap {
            scalars.append(self.table[element].scalarized())
        }

        //let gammaTensor = Tensorf(stacking: scalars).reshaped(to: input.semanticMap.shape)
        let gammaTensor = Tensorf(shape: input.semanticMap.shape, scalars: scalars)
        
        return input.input * gammaTensor
    }
```

Error: 
```
Fatal error: Incompatible shapes: [1,256,256,12] vs. [65536]: file /swift-base/tensorflow-swift-apis/Sources/TensorFlow/Bindings/EagerExecution.swift, line 301
Current stack trace:
0    libswiftCore.so                    0x00007ff5c0e79910 swift_reportError + 50
1    libswiftCore.so                    0x00007ff5c0eeab90 _swift_stdlib_reportFatalErrorInFile + 115
2    libswiftCore.so                    0x00007ff5c0bcae31 <unavailable> + 1461809
3    libswiftCore.so                    0x00007ff5c0bcaa77 <unavailable> + 1460855
4    libswiftCore.so                    0x00007ff5c0bcb012 <unavailable> + 1462290
5    libswiftCore.so                    0x00007ff5c0bc9460 _assertionFailure(_:_:file:line:flags:) + 517
6    libswiftTensorFlow.so              0x00007ff5bd72c61a <unavailable> + 4367898
7    libswiftTensorFlow.so              0x00007ff5bd72c620 checkOk(_:file:line:) + 436
8    libswiftTensorFlow.so              0x00007ff5bd4e0b80 TFE_Op.evaluateUnsafe() + 533
9    libswiftTensorFlow.so              0x00007ff5bd4e0e70 TFE_Op.execute<A>(_:) + 143
10   libswiftTensorFlow.so              0x00007ff5bd4f5442 <unavailable> + 2044994
Current stack trace:
	frame #11: 0x00007ff57c049673 $__lldb_expr26`AD__$s14__lldb_expr_255CLADEV14callAsFunctiony10TensorFlow0G0VySfGAC5InputVF__vjp_src_0_wrt_1(input=<unavailable>, self=<unavailable>) at <Cell 3>:35:28
	frame #12: 0x00007ff57c04a09a $__lldb_expr26`AD__$s14__lldb_expr_2510TensorFlow0C0VySfGAA5CLADEVcfU___vjp_src_0_wrt_0(c=<unavailable>) at <Cell 3>:45:22
	frame #21: 0x00007ff57c0446da $__lldb_expr26`main at <Cell 3>:44
```

**Comment by marcrasi** [comment]
> I tried running the code on remote Linux server with CUDA 10.1 and v0.8 toolchain and it really appears to work. Unfortunately this is veeeeery slow, I will try to look at it

The slowness is probably caused by the array indexing `self.table[element]`. Currently, derivatives of array indexing are very slow. It's something we want to fix eventually. (See this email on the mailing list for some explanation if you're interested: https://groups.google.com/a/tensorflow.org/g/swift/c/az0-tV15_rA/m/lTOvkO2RAAAJ).

Embedding shouldn't suffer from the same problem, so it should be faster.

> I've also tried to optimize stuff by using Floats and getting really weird error on Linux

Yes that error is really weird. It made me suspicious that the derivative for `init(shape:scalars:)` defined in swift-apis might be wrong. [So I took a look and indeed it is wrong!](https://github.com/tensorflow/swift-apis/blob/51b0bd126476b64543676268774e5691ad5176a5/Sources/TensorFlow/Core/Tensor.swift#L431) I'll send a PR fixing it soon. In the meantime, your workaround with `.reshaped` is good.

**Comment by s1ddok** [comment]
@marcrasi thanks for taking a look at it! Should we repurpose this issue to the fact that same code on macOS and linux work in different ways? Or just close it for now?

**Comment by marcrasi** [comment]
I have filed a separate issue for the macos crash, so that it's easy to read it without being distracted by all the other discussion.

Let's close this issue if you're happy with everything else.

**Comment by s1ddok** [comment]
Yes, thanks a lot!
